### ISSUE TITLE
Improve concurrency

### ISSUE THREAD

**Issue by jeffkaufman** [gwillen/solstice-audio-test#43]
Today our architecture looks like:

1. Requests come in to Nginx, which handles HTTPS

2. Nginx reverse proxies to a long-running `http.server` / `BaseHTTPRequestHandler`

This has two main problems:

* We can't take advantage of any concurrency that the server might be able to provide

* We're getting freezes (#42)

Instead, I think we should be doing something more like:

1. Requests come in to Nginx, which handles HTTPS

2. Nginx uses the wsgi protocol to delegate to a uwsgi server running all the components of the app that do not need global state, primarily opus decoding and encoding. 

3. For handling global state we run a singleton process, communication over UNIX socket or something.

**Comment by jeffkaufman** [comment]
One thing that makes this tricky is that part of our global state is each user's opus state, and I don't know if that can be serialized?

**Comment by jeffkaufman** [comment]
Since every user has equal processing needs, it would be possible to have users divided up by process.  Then opus state never needs to be transferred. The main downside of this is that it's not a very normal way of handling requests, and so any sort of pre-existing tooling is going to be a poor fit. I'd like to avoid this if we can.

**Comment by jeffkaufman** [comment]
There are two very different kinds of global state:

* The buffer

* Information about which users are in the system, what their names are, what positions they are etc.

These don't even need to be handled by the same server; they are totally independent

**Comment by dspeyer** [comment]
I expect metainfo is cheap and doesn't need much cleverness.

ISTR there's a way to use shared memory with numpy.  Clunky, but this would be a simple case of it.

**Comment by jeffkaufman** [comment]
I think something like https://research.wmz.ninja/articles/2018/03/on-sharing-large-arrays-when-using-pythons-multiprocessing.html should work

uwsgi in prefork+threading mode should let us make the `multiprocessing.Array` before forking and then use it everywhere

**Comment by jeffkaufman** [comment]
I now have it working over wsgi, and am having a go at converting all global state to use multiprocessing.Value and multiprocessing.Array

**Comment by jeffkaufman** [comment]
There is quite a bit of state that doesn't really need to be global but is read on every request.  I'm not planning to fix any of this initially, so at least initially there's going to be a lot of awkward lock thrashing

**Comment by jeffkaufman** [comment]
I'm stuck on what to do with the opus decoder state.  It's a pointer, which means if we create a new one in a child process then even if we write it a shared memory the other child process will not be able to read it.  I thought maybe we could preallocate a whole bunch, and flush them when we're not using them, but I think the copy-on-write semantics of forking means that won't work?

Using threads instead will not let us take advantage of multiple CPUs.

There doesn't seem to be a way to ask the opus library to serialize its state.

The main other option that comes to mind is to handle forking ourselves.  We tell uwsgi that it should run only one process, And we use multiprocessing to start up a bunch of decoder/encoder processes. We can track which user's state is in which process, and have those processes do the hard work

**Comment by gwillen** [comment]
> Using threads instead will not let us take advantage of multiple CPUs.

This is true for native python but not for wrapped C libraries. (I just did a simple test with 8 threads doing matrix multiplication using numpy; I can easily get python reporting 500+% CPU usage, i.e. using multiple full cores.) I expect that both opus and numpy will happily take advantage of multiple CPUs in our setup, merely for switching on threading in the HTTP server.

**Comment by jeffkaufman** [comment]
I have now switched us over to WSGI

**Comment by jeffkaufman** [comment]
Before we can enable multiple threads and we're going to need to think hard about how we are going to handle global state.  A good first step would be to set up locking for all of our read modify write operations

**Comment by jeffkaufman** [comment]
I tried a very crude test: jefftk.com is running on a two-CPU server, so I tried passing `--threads=2` to uwsgi.  I saw uwsgi go from using 100% cpu to 150%.  I don't have an exact number of clients it could handle, but before threading it could handle 50 but not 75, and turning on threading did not change this.

**Comment by jeffkaufman** [comment]
We should run some profiling to figure out how many users we could handle in one thread if we didn't have opus encoding and decoding. If that is high enough we could have a structure like:

```
[request] ->
   [many opus decoders, sharded on userid] ->
   [existing app] ->
   [many opus encoders, sharded on userid] ->
   [response]
```

**Comment by jeffkaufman** [comment]
Duplicate of https://github.com/gwillen/solstice-audio-test/issues/41
